Care-O-Bot4 (Cob4) test and visualization scripts

juandhv opened the pull request:
Care-O-Bot4 (Cob4) is a dual-arm service robot. This PR includes two robowflex scripts to test and visualize motion planning queries for the Cob4. The corresponding robowflex resources have also been [added](https://github.com/KavrakiLab/robowflex_resources/pull/14).

zkingston commented on robowflex_library/include/robowflex_library/detail/cob4.h:
```diff
@@ -0,0 +1,130 @@
+    /** \class robowflex::Cob4RobotConstPtr
+        \brief A const shared pointer wrapper for robowflex::Cob4Robot. */
+
+    /** \brief Convenience class that describes the default setup for Cob4.
```
Cob4 -> Care-O-Bot4?

juandhv replied:
Done!

zkingston commented on robowflex_library/include/robowflex_library/detail/cob4.h:
```diff
@@ -0,0 +1,130 @@
+         */
+        bool initialize();
+
+        /** \brief Inserts the caster links if they don't exist.
+         *  \param[in] doc urdf description to be processed.
+         *  \return True on success.
+         */
+        bool addCastersURDF(tinyxml2::XMLDocument &doc);
```
Does this robot also have the exact same casters as the Fetch?

juandhv replied:
You're right. I left this when I used the Fetch specification to implement the one for the Care-O-Bot4. I've cleaned it up.

zkingston commented on robowflex_library/scripts/cob4_test.cpp:
```diff
@@ -0,0 +1,105 @@
+    ROS ros(argc, argv);
+
+    // Create the Care-O-Bot4 robot.
+    auto cob4 = std::make_shared<Robot>("cob4");
+    cob4->initialize("package://robowflex_resources/cob/robots/cob4-8.urdf.xacro",  // urdf
+                     "package://robowflex_resources/cob/config/cob4-8.srdf",        // srdf
+                     "package://robowflex_resources/cob/config/joint_limits.yaml",  // joint limits
+                     "package://robowflex_resources/cob/config/kinematics.yaml"     // kinematics
+    );
```
Why not use the detail implementation?

juandhv replied:
Forgot to do this after creating the detail implementation. Done!

zkingston commented on robowflex_library/scripts/cob4_test.cpp:
```diff
@@ -0,0 +1,105 @@
+    cob4->setGroupState(RIGHT_ARM, {0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0});
+    cob4->setGroupState(LEFT_ARM, {0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0});
+
+    // Create the default planner for the COB4.
+    auto planner = std::make_shared<OMPL::OMPLPipelinePlanner>(cob4);
```
Same here, why not use the detail planner?

zkingston commented on robowflex_library/scripts/cob4_visualization.cpp:
```diff
@@ -0,0 +1,166 @@
+    auto cob4 = std::make_shared<Cob4Robot>();
+    cob4->initialize();
+
+    // Alternatively:
+    //    auto cob4 = std::make_shared<Robot>("cob4");
+    //    cob4->initialize("package://robowflex_resources/cob/robots/cob4-8.urdf.xacro",  // urdf
+    //                     "package://robowflex_resources/cob/config/cob4-8.srdf",        // srdf
+    //                     "package://robowflex_resources/cob/config/joint_limits.yaml",  // joint limits
+    //                     "package://robowflex_resources/cob/config/kinematics.yaml"     // kinematics
+    //    );
```
I'd just remove this comment.

juandhv replied:
Done!

zkingston commented on robowflex_library/scripts/cob4_visualization.cpp:
```diff
@@ -0,0 +1,166 @@
+    rviz.updateScene(scene);
+
+    // Create the default planner for the COB4.
+    auto planner = std::make_shared<OMPL::OMPLPipelinePlanner>(cob4);
```
The detail planner?

zkingston commented on robowflex_library/scripts/cob4_visualization.cpp:
```diff
@@ -0,0 +1,166 @@
+    request_right_arm.setStartConfiguration(cob4->getScratchState());
+
+    // Create a motion planning request with a pose goal for the right arm.
+    RobotPose goal_pose_right = RobotPose::Identity();
+    goal_pose_right.translate(Eigen::Vector3d{0.6, -0.26, 0.95});
+    goal_pose_right.rotate(Eigen::Quaterniond{0.0, 1.0, 0.0, 0.0});
+    request_right_arm.setGoalPose(RIGHT_EE, "base_link", goal_pose_right);
```
Use the TF::createPose* functions to clean this up.

juandhv replied:
Done! Thanks for the suggestion!

zkingston commented on robowflex_library/scripts/cob4_visualization.cpp:
```diff
@@ -0,0 +1,166 @@
+    rviz.updateTrajectory(res);
+
+    // Visualize resulting state.
+    trajectory_msgs::JointTrajectoryPoint point;
+    res.trajectory_->getLastWayPoint().copyJointGroupPositions(
+        res.trajectory_->getLastWayPoint().getJointModelGroup(RIGHT_ARM), point.positions);
+    cob4->setGroupState(RIGHT_ARM, std::vector<double>(point.positions));
```
Use the robowflex::Trajectory functions to get the last point

juandhv replied:
Please take a look to the new version, and let me know if that is what you meant.

zkingston commented on robowflex_library/scripts/cob4_visualization.cpp:
```diff
@@ -0,0 +1,166 @@
+    request_left_arm.setStartConfiguration(cob4->getScratchState());
+
+    // Create a motion planning request with a pose goal for the left arm.
+    RobotPose goal_pose_left = RobotPose::Identity();
+    goal_pose_left.translate(Eigen::Vector3d{0.38, 0.26, 0.76});
+    goal_pose_left.rotate(Eigen::Quaterniond{0.707, 0.0, 0.707, 0.0});
+    request_left_arm.setGoalPose(LEFT_EE, "base_link", goal_pose_left);
```
Use the TF::createPose* functions to clean this up.

juandhv replied:
Done!

zkingston commented on robowflex_library/scripts/cob4_visualization.cpp:
```diff
@@ -0,0 +1,166 @@
+    rviz.updateTrajectory(res);
+
+    // Visualize resulting state.
+    res.trajectory_->getLastWayPoint().copyJointGroupPositions(
+        res.trajectory_->getLastWayPoint().getJointModelGroup(LEFT_ARM), point.positions);
+    cob4->setGroupState(LEFT_ARM, std::vector<double>(point.positions));
```
Same as above.

juandhv replied:
Done!

zkingston commented on robowflex_library/src/detail/cob4.cpp:
```diff
@@ -0,0 +1,181 @@
+
+bool Cob4Robot::addCastersURDF(tinyxml2::XMLDocument &doc)
+{
+    for (const auto name : {"bl_caster", "br_caster", "fl_caster", "fr_caster"})
```
This should probably be changed for the Cob4? I doubt its the exact same as the fetch.

juandhv replied:
Done!

zkingston reviewed (changes requested):
Some nits.

ChamzasKonstantinos commented on robowflex_library/yaml/test_cob4.yml:
```diff
@@ -0,0 +1,41 @@
+robot_model_name: cob4
+robot_state:
+  joint_state: 
+    position: [0, 0, 0.05, 0, 0, 0, 1.32, 1.4, -0.2, 1.72, 0.0, 1.66, 0.0, 0.04, 0.04]    
```
The size of this vector is different from the size of the joints (one is 17 one is 15)

juandhv replied:
Done!

ChamzasKonstantinos commented on robowflex_library/yaml/test_cob4.yml:
```diff
@@ -0,0 +1,41 @@
+    position: [0, 0, 0.05, 0, 0, 0, 1.32, 1.4, -0.2, 1.72, 0.0, 1.66, 0.0, 0.04, 0.04]    
+    name: [arm_left_1_joint, arm_left_2_joint, arm_left_3_joint, arm_left_4_joint, arm_left_5_joint, arm_left_6_joint, arm_left_7_joint, arm_right_1_joint, arm_right_2_joint, arm_right_3_joint, arm_right_4_joint, arm_right_5_joint, arm_right_6_joint, arm_right_7_joint, head_1_joint, head_2_joint, head_3_joint]
+fixed_frame_transforms:
+  - child_frame_id: world
```
This should be base_footprint, since that is the base_frame of the robot

juandhv replied:
Done!

ChamzasKonstantinos commented on robowflex_library/yaml/test_cob4.yml:
```diff
@@ -0,0 +1,41 @@
+name: (noname)
+robot_model_name: cob4
```
This should be cob4-8 since that is the name in the srdf

juandhv replied:
Done!

ChamzasKonstantinos commented on robowflex_library/yaml/test_cob4.yml:
```diff
@@ -0,0 +1,41 @@
+name: (noname)
```
This entry (the robot state) throws warnings errors in Rviz. You can either fix the errors or you can delete the robot_state in the scene completely it is not needed or used.

juandhv replied:
done!

ChamzasKonstantinos commented:
![temp](https://user-images.githubusercontent.com/5930462/129907321-cdfa8337-2c33-4ad7-b262-53896ac8a77e.png)
I run the RviZ script and works as expected. 
A couple of notes/suggestions from my side as well.
when you load the model in RviZ the following errors/warnings appear: 
```
[ WARN] [1629293465.992645078]: Setting the scene for model 'cob4' but model 'cob4-8' is loaded.
[ERROR] [1629293465.992708625]: Given transform is to frame 'world', but frame 'base_footprint' was expected.
[ERROR] [1629293465.992752469]: Different number of names and positions in JointState message: 17, 15
```
This can be fixed by fixing the entry on the cob4_test.yml (see comments) 

Also you might want to consider adding a both_arms group to perform double arm manipulation with a single planner similar to baxter see here: 
https://github.com/KavrakiLab/robowflex_resources/blob/master/baxter/config/ompl_planning.yaml#L130
https://github.com/KavrakiLab/robowflex_resources/blob/master/baxter/config/baxter.srdf#L68
https://github.com/KavrakiLab/robowflex/blob/master/robowflex_library/scripts/baxter_multi_target.cpp

juandhv commented:
@ChamzasKonstantinos I've addressed your comments, thanks! I will have the both_arms group in a separate PR, because I will need to make some changes in robowflex_resources.

zkingston reviewed (approved):
LGTM
